### PR TITLE
mark `min_exhaustive_patterns` as complete

### PR DESCRIPTION
**Pull request by Nadrieril**
This is step 1 and 2 of my [proposal](https://github.com/rust-lang/rust/issues/119612#issuecomment-1918097361) to move `min_exhaustive_patterns` forward. The vast majority of in-tree use cases of `exhaustive_patterns` are covered by `min_exhaustive_patterns`. There are a few cases that still require `exhaustive_patterns` in tests and they're all behind references.

r? @ghost

### CODE REVIEW

File: compiler/rustc_mir_build/src/build/matches/simplify.rs [hunk 1]
```diff
@@ -271,7 +271,8 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
             PatKind::Variant { adt_def, args, variant_index, ref subpatterns } => {
                 let irrefutable = adt_def.variants().iter_enumerated().all(|(i, v)| {
                     i == variant_index || {
-                        self.tcx.features().exhaustive_patterns
+                        (self.tcx.features().exhaustive_patterns
```
Comment (compiler-errors) [inline compiler/rustc_mir_build/src/build/matches/simplify.rs:274]: what does this do specifically?

Comment (Nadrieril) [reply compiler/rustc_mir_build/src/build/matches/simplify.rs:274]: The following: if the pattern is `SomeEnum::Variant3` and all the other variants are uninhabited, then we skip the discriminant test. I don't fully understand why we only do this when the feature is on, my guess is someone was overly-conservative.

Comment (Nadrieril) [reply compiler/rustc_mir_build/src/build/matches/simplify.rs:274]: See this: https://github.com/rust-lang/rust/blob/master/tests/ui/closures/2229_closure_analysis/run_pass/multivariant.rs

**Comment by compiler-errors** [comment]
If you pull 51ada89d9d2d99211e1acf83f09cb2ece532be63, b9a1b41, e1df69a into a separate PR, I can approve that immediately. There's nobody you need to ask permission to use a feature in the compiler except for someone from the compiler, so that doesn't need T-lang approval or anything

I would recommend then 0e76047 landing separately with approval from T-libs *after* that, along with pointing them to the relevant tracking issues or anything that would increase their confidence that this is a sound subset of the exhaustive patterns feature.

As for 464d5e7, I wonder if all of the modified tests should instead be made into ones that have two revisions between `min` and `full`.

**Comment by Nadrieril** [comment]
Thank you! I'm thinking I do need approval for marking it as complete because it's a requirement of the T-lang ["experimental feature"](https://github.com/rust-lang/lang-team/blob/master/src/how_to/experiment.md) process I've been using: "This flag has to stay until an RFC is accepted, even if the implementation is in good shape".

**Comment by compiler-errors** [comment]
oh huh, didn't know that

**Comment by Nadrieril** [comment]
I've essentially been side-stepping T-lang to make progress on this (mostly for the never patterns stuff tbh), so I gotta check back with them at some point

**Comment by Nadrieril** [comment]
I've split off what doesn't require T-lang over there: https://github.com/rust-lang/rust/pull/120775

**Review by compiler-errors — APPROVED** [review]
r=me when the FCP is finished

**Pull request merged**